mirrord-agent/src/tcp/outgoing.rs

```diff
@@ -0,0 +1,177 @@
+use std::{collections::HashMap, path::PathBuf, thread};
+
+use mirrord_protocol::{tcp::outgoing::*, ConnectionId, ResponseError};
+use tokio::{
+    io::AsyncWriteExt,
+    net::{
+        tcp::{OwnedReadHalf, OwnedWriteHalf},
+        TcpStream,
+    },
+    select,
+    sync::mpsc::{self, Receiver, Sender},
+};
+use tokio_stream::{StreamExt, StreamMap};
+use tokio_util::io::ReaderStream;
+use tracing::{trace, warn};
+
+use crate::{error::AgentError, runtime::set_namespace, util::run_thread};
+
+type Request = TcpOutgoingRequest;
+type Response = TcpOutgoingResponse;
+
+/// Handles (briefly) the `TcpOutgoingRequest` and `TcpOutgoingResponse` messages, mostly the
+/// passing of these messages to the `interceptor_task` thread.
+pub(crate) struct TcpOutgoingApi {
+    /// Holds the `interceptor_task`.
+    _task: thread::JoinHandle<Result<(), AgentError>>,
+
+    /// Sends the `Request` to the `interceptor_task`.
+    request_tx: Sender<Request>,
+
+    /// Reads the `Response` from the `interceptor_task`.
+    response_rx: Receiver<Response>,
+}
+
+impl TcpOutgoingApi {
+    pub(crate) fn new(pid: Option<u64>) -> Self {
+        let (request_tx, request_rx) = mpsc::channel(1000);
+        let (response_tx, response_rx) = mpsc::channel(1000);
+
+        let task = run_thread(Self::interceptor_task(pid, request_rx, response_tx));
+
+        Self {
+            _task: task,
+            request_tx,
+            response_rx,
+        }
+    }
+
+    /// Does the actual work for `Request`s and prepares the `Responses:
+    async fn interceptor_task(
+        pid: Option<u64>,
+        mut request_rx: Receiver<Request>,
+        response_tx: Sender<Response>,
+    ) -> Result<(), AgentError> {
+        if let Some(pid) = pid {
+            let namespace = PathBuf::from("/proc")
+                .join(PathBuf::from(pid.to_string()))
+                .join(PathBuf::from("ns/net"));
+
+            set_namespace(namespace).unwrap();
+        }
+
+        let mut writers: HashMap<ConnectionId, OwnedWriteHalf> = HashMap::default();
+        let mut readers: StreamMap<ConnectionId, ReaderStream<OwnedReadHalf>> =
+            StreamMap::default();
+
+        loop {
+            select! {
+                biased;
+
+                // [layer] -> [agent]
+                request = request_rx.recv() => {
+                    trace!("interceptor_task -> request {:?}", request);
+
+                    match request {
+                        Some(request) => {
+                            match request {
+                                // [user] -> [layer] -> [agent] -> [layer]
+                                // `user` is asking us to connect to some remote host.
+                                TcpOutgoingRequest::Connect(ConnectRequest { remote_address }) => {
```

Comment:
I think the Request/Response terminology might be confusing (of the enum, `ConnectRequest` is fine). I'd call it `LayerTcpOutgoing` and `DaemonTcpOutgoing` or smth like that?

Reply:
Sure, I wanted to unify terminology for layer -> agent, and agent -> layer messages. As I was using this request/response for the features I've been implementing, so I went ahead and used the same here.

If we're going to change it to be layer/daemon instead of request/response, it's fine for me, but we should change it everywhere as well.

Reply:
Changed it to be Layer/Daemon, instead of Request/Response.